test/model/test/model_transaction/main.cpp

```diff
@@ -748,6 +748,36 @@ test_transaction_prepared(void)
     txn1->prepare(60);
     txn1->rollback();
     testutil_assert(table->get(key1) == value1);
+
+    /* Overlapping transactions with prepare, special read behaviors (1-3) */
+    /* Read behavior 1: insert key3 (txn1), begin txn2, prepare txn1, txn2 can't find key3. */
+    txn1 = database.begin_transaction();
+    testutil_check(table->insert(txn1, key3, value3));
+    txn2 = database.begin_transaction();
+    txn1->prepare(70);
+    testutil_assert(table->get_ext(txn2, key3, v) == WT_NOTFOUND);
+    txn1->commit(70, 70);
+    txn2->commit(70);
+
+    /* Read behavior 2: insert key3 (txn1), prepare txn1, begin txn2, txn2 has read conflict. */
+    txn1 = database.begin_transaction();
+    testutil_check(table->insert(txn1, key3, value1));
+    txn1->prepare(80);
+    txn2 = database.begin_transaction();
+    testutil_assert(table->get_ext(txn2, key3, v) == WT_PREPARE_CONFLICT);
+    txn1->commit(80, 80);
+    txn2->commit(80);
+
+    /* Read behavior 3: insert key3 (txn1), prepare txn1, begin txn2, commit txn1, txn2 can see
+     * txn1's update. */
```

Comment:
This is surprising to me. Am I understanding correctly that under snapshot isolation, there are scenarios where data could change underneath us?

Or is it the case that we read after the commit (as opposed to before), so it's OK because we didn't technically "see" the value change underneath us.

Reply:
Good question, I also thought this. From my initial investigation, this is intended behavior and I found some Python tests that exercise it (`test_prepare_cursor01.py` at `test_cursor_navigate_prepare_transaction`).

To play advocate for why this might make sense, if the transaction did read while the value was prepared and not committed, it would have already rolled back, so the only transaction that can survive this scenario must have only witnessed the post- value.

Reply:
> To play advocate for why this might make sense, if the transaction did read while the value was prepared and not committed, it would have already rolled back, so the only transaction that can survive this scenario must have only witnessed the post- value.

Yeah, I think that's right. It seems like the other side of the coin to how, if we don't read it before the commit, we never saw the value change.